### ISSUE TITLE
[help] how to save model with given id

### ISSUE THREAD

**Issue by zzswang** [maritz/nohm#140]
## Requirement

We need to save a model with given id.

## Question

What schema should I defined? And when save a mode, should I add id property.

Currently my code is:

```
// the schema
const { Nohm } = require("nohm");

Nohm.setPrefix("bus");
const Pet = Nohm.model("Pet", {
  properties: {
    id: { type: "string", index: true, unique: true }, // vin
    ns: { type: "string", index: true }
  }
});

const pet = new Pet();

 // set some properties
 pet.property({
      id: "a-given-id-not-auto-generated",
      ns: "/Gondor",
 });
```

We find that model will be stored with an extra auto generated uuid in the key.

like

`nohm:hash:Pet:c8792349-sdf234234-dsfs-234ssdfsf-234234`

And when we load it

```
Pet.load("a-given-id-not-auto-generated");
```

it throw not found error

**Comment by maritz** [comment]
I'm not exactly sure what you mean.

### Option 1 
You want some other method than uuid or incremental ids for all your ids for a model.

This is covered by [idGenerators](https://maritz.github.io/nohm/#id-generation).

### Option 2

You have some specific (and rare) cases where you don't want the default generation to happen when creating new instances and you already have an ID that you want to use for it and you are either sure it doesn't already exist or you are fine with updating an old one if it does exist.

```javascript
const Pet = Nohm.model("Pet", {
  properties: {
    ns: { type: "string", index: true }
  }
});

const pet = new Pet();

pet.id = 'foo-bar';
pet.property({
  ns: "/Gondor",
});

await pet.save();

await Pet.load('foo-bar'); // should return the pet we just created.

// HOWEVER!
// if you then do this:

const pet2 = new Pet();

pet2.id = 'foo-bar';
pet2.property({
  ns: "/Mordor",
});

await pet2.save(); // this UPDATES the "pet1" we created earlier, because the id is already in the database.
```

Again: **WARNING!** You are completely responsible for the ids this way and have to make sure that when you create a new instance the id you pass is available at the time of saving the object!
For example the built-in [incremental id generator](https://github.com/maritz/nohm/blob/master/ts/idGenerators.ts#L21) increments before saving. Thus no 2 saves can ever have the same id, no matter how concurrent. It has the downside that if something goes very wrong, it may skip numbers, but that's more acceptable than accidentally updating when you want to create.

If you don't set the id before saving a new instance, it will fall back to creating a uuid.

**Comment by zzswang** [comment]
@maritz 

Thank you. option 2 is exactly what we need.

We use Redis to save some data which comes from other services. These modes will have ids, and  are used in api call, like `/pets/:petId`.

**Comment by fedecaccia** [comment]
Thank you guys for your clarifications on this point!

**Comment by maritz** [comment]
@fedecaccia Keep in mind that there is a bug (#141) with setting your own id. I currently don't really have the time to deal with it, sadly.

**Comment by maritz** [comment]
@fedecaccia @zzswang Sorry about the long delay.  
Actually having thought about it more I think this is the wrong approach.

Wouldn't it be simpler to have an extra property "externalId" that you populate with your custom ids and then just let nohm do uuid stuff internally?!

```javascript

const Pet = Nohm.model("Pet", {
  properties: {
    ns: { type: "string", index: true }
    external_id: { type: "string", unique: true }
  }
});

const pet1 = new Pet();

pet1.property({
  ns: '/Gondor',
  external_id: 'foo-bar'
});

await pet1.save();

const pets = await Pet.findAndLoad({external_id: 'foo-bar'}); // should return the pet we just created. 
// Since it's a unique the find isn't really slower than a load.
// The only difference is that findAndLoad returns an array of found pets and doesn't throw if none are found
const pet = pets[0];
```

You should probably make a helper method that encapsulates this loading to make it simpler.

```javascript
// To update something based on its externalId load it via find as well and edit it:

const loadByExternalId = async (id) => {
  const pets = await Pet.findAndLoad({external_id: id});
  // mimic Pet.load() behaviour of throwing a 'not found' if none are available with that id
  if (pets.length === 0) {
    throw new Error('not found');
  }
  return pets[0];
};

const pet2 = await loadByExternalId('foo-bar');

pet2.property({
  ns: "/Mordor",
});

await pet2.save(); // this UPDATES the "pet1" we created earlier
```

I think in nohm v3 I'll probably make id internal and only provide a getId() to read it. That should prevent a lot of confusion.

Any feedback or better ideas for this are very welcome!

**Comment by tarekch** [comment]
Hello,
I know this has been a long time but i have a workaround for the manual id problem.
I absolutely needed to have my own manual ids, and the idea proposed to have an `externalId` doesnt really suit me as i will still need to do `.find()` using multiple primary keys which is not supported if i used `externalId`

To make a story short, i noticed that you call the idGenerator bound to `this` so i simply created this idGenerator:
```
return function idGenerator() {
    return _.get(this.allProperties(), 'my_own_id') || shortid.generate()
 }
```

Basically i do have the property `my_own_id` on the model which of course is a duplicate of `id`. This will only be a problem if `my_own_id` gets updated afterwards but you can create a wrapper to protect from that.


Hope this helps
